File: 2020-08-month-long-challenge/day21.py
#!/usr/bin/env python3

# Day 21: Sort Array By Parity
#
# Given an array A of non-negative integers, return an array consisting of all
# the even elements of A, followed by all the odd elements of A.
# You may return any answer array that satisfies this condition.
#
# Note:
# - 1 <= A.length <= 5000
# - 0 <= A[i] <= 5000

class Solution:
    def sortArrayByParity(self, A: [int]) -> [int]:
        # One-liner solution
        # return sorted(A, key=lambda x: x % 2)
        
        # Faster solution, as this is a simple partition
        # evens = []
        # odds = []
        # for number in A:
        #     if number % 2 == 0:
        #         evens.append(number)
        #     else:
        #         odds.append(number)
        # return evens + odds

        # Even faster solution, thanks to Python list comprehensions
        # (runtime beats 99% of submissions, memory usage beats 93%)
        return [n for n in A if n % 2 == 0] + [n for n in A if n % 2 != 0]

# Tests
assert Solution().sortArrayByParity([]) == []
assert Solution().sortArrayByParity([1,2]) == [2,1]
assert Solution().sortArrayByParity([3,1,2,4]) in [[2,4,3,1], [4,2,3,1], [2,4,1,3], [4,2,1,3]]

File: 2020-08-month-long-challenge/day21.js
#!/usr/bin/env nodejs

/* Day 21: Sort Array By Parity
 *
 * Given an array A of non-negative integers, return an array consisting of all
 * the even elements of A, followed by all the odd elements of A.
 * You may return any answer array that satisfies this condition.
 *
 * Note:
 * - 1 <= A.length <= 5000
 * - 0 <= A[i] <= 5000
 */

/**
 * @param {number[]} A
 * @return {number[]}
 */
var sortArrayByParity = function(A) {
    let evens = []
    let odds = []
    for (number of A) {
        if (number % 2 == 0) {
            evens.push(number)
        } else {
            odds.push(number)
        }
    }
    return evens.concat(odds)
};

// Let's build a list comparison function since [a,b,c] == [d,e,f] doesn't work
// in JavaScript
const eq = x => y => x == y
const range = n => [...Array(n).keys()]
const map = fn => xs => xs.map(fn)
const zipwith = fn => xs => ys =>
    map(i => fn(xs[i])(ys[i]))(range(Math.min(xs.length, ys.length)))
const all = xs => xs.every(x => x == true)
const list_equal = xs => ys => eq(xs.length)(ys.length)
    && all(zipwith(eq)(xs)(ys))

// Tests
console.assert(list_equal(sortArrayByParity([]))([]))
console.assert(list_equal(sortArrayByParity([1,2]))([2,1]))
console.assert(list_equal(sortArrayByParity([3,1,2,4]))([2,4,3,1]))
